src/substrait/function_registry.py

```diff
@@ -0,0 +1,378 @@
+from substrait.gen.proto.type_pb2 import Type
+from importlib.resources import files as importlib_files
+import itertools
+from collections import defaultdict
+from collections.abc import Mapping
+from pathlib import Path
+from typing import Any, Optional, Union
+from .derivation_expression import evaluate
+
+import yaml
+import re
+
+
+DEFAULT_URI_PREFIX = "https://github.com/substrait-io/substrait/blob/main/extensions"
+
+
+# mapping from argument types to shortened signature names: https://substrait.io/extensions/#function-signature-compound-names
+_normalized_key_names = {
+    "i8": "i8",
+    "i16": "i16",
+    "i32": "i32",
+    "i64": "i64",
+    "fp32": "fp32",
+    "fp64": "fp64",
+    "string": "str",
+    "binary": "vbin",
+    "boolean": "bool",
+    "timestamp": "ts",
+    "timestamp_tz": "tstz",
+    "date": "date",
+    "time": "time",
+    "interval_year": "iyear",
+    "interval_day": "iday",
+    "interval_compound": "icompound",
+    "uuid": "uuid",
+    "fixedchar": "fchar",
+    "varchar": "vchar",
+    "fixedbinary": "fbin",
+    "decimal": "dec",
+    "precision_time": "pt",
+    "precision_timestamp": "pts",
+    "precision_timestamp_tz": "ptstz",
+    "struct": "struct",
+    "list": "list",
+    "map": "map",
+}
+
+
+def normalize_substrait_type_names(typ: str) -> str:
+    # Strip type specifiers
+    typ = typ.split("<")[0]
+    # First strip nullability marker
+    typ = typ.strip("?").lower()
+
+    if typ.startswith("any"):
+        return "any"
+    elif typ.startswith("u!"):
+        return typ
+    else:
+        return _normalized_key_names[typ]
+
+
+def to_integer_option(txt: str):
+    if txt.isnumeric():
+        return ParameterizedType.IntegerOption(literal=int(txt))
+    else:
+        return ParameterizedType.IntegerOption(
+            parameter=ParameterizedType.IntegerParameter(name=txt)
+        )
+
+
+# TODO try using antlr grammar here as well
+def to_parameterized_type(dtype: str):
+    if "?" in dtype:
+        dtype = dtype.replace("?", "")
+        nullability = Type.NULLABILITY_NULLABLE
+    else:
+        nullability = Type.NULLABILITY_REQUIRED
+
+    if dtype == "boolean":
+        return ParameterizedType(bool=Type.Boolean(nullability=nullability))
+    elif dtype == "i8":
+        return ParameterizedType(i8=Type.I8(nullability=nullability))
+    elif dtype == "i16":
+        return ParameterizedType(i16=Type.I16(nullability=nullability))
+    elif dtype == "i32":
+        return ParameterizedType(i32=Type.I32(nullability=nullability))
+    elif dtype == "i64":
+        return ParameterizedType(i64=Type.I64(nullability=nullability))
+    elif dtype == "fp32":
+        return ParameterizedType(fp32=Type.FP32(nullability=nullability))
+    elif dtype == "fp64":
+        return ParameterizedType(fp64=Type.FP64(nullability=nullability))
+    elif dtype == "timestamp":
+        return ParameterizedType(timestamp=Type.Timestamp(nullability=nullability))
+    elif dtype == "timestamp_tz":
```

Comment:
Consider creating an enum/type/constant for the types in the future so that there is only one place where the names are.  (Not this PR)

Reply:
Got rid of this part of the code while rewriting with antlr, but will take it into consideration.